### PR TITLE
[subset] support option: --layout-features

### PR DESCRIPTION
**Pull request by qxliu76**

### CODE REVIEW

File: src/hb-ot-layout.h [hunk 1]
```diff
@@ -312,15 +312,17 @@ hb_ot_layout_collect_features (hb_face_t      *face,
 			       hb_tag_t        table_tag,
 			       const hb_tag_t *scripts,
 			       const hb_tag_t *languages,
-			       const hb_tag_t *features,
+			       const hb_set_t *features,
```
Comment (garretrieger) [inline src/hb-ot-layout.h:315]: We can't change the signature of an existing API method. Instead I would add a second version method that takes features as a set. Then that could internally convert features to an array and call this method, or you could update this method to convert the array to a set and call the new one.

Additionally I'd drop the retain_all_features bool and instead stick with the current behaviour that if the features set is a nullptr that means keep all.

Comment (khaledhosny) [reply src/hb-ot-layout.h:315]: > if the features set is a nullptr that means keep all.

What if one wants to drop all features?

Comment (garretrieger) [reply src/hb-ot-layout.h:315]: Passing an empty set would drop all features.

Comment (khaledhosny) [reply src/hb-ot-layout.h:315]: Right.

File: src/hb-subset-plan.cc [hunk 2]
```diff
@@ -68,20 +68,95 @@ _remap_indexes (const hb_set_t *indexes,
...
+				lookup_indices);
+
+  free (features);
+}
+
+static void _collect_subset_features (hb_face_t      *face,
+				      hb_tag_t        table_tag,
+				      const hb_set_t *layout_features_to_retain,
+				      bool            retain_all_features,
+				      hb_set_t       *lookup_indices /* OUT */)
```
Comment (garretrieger) [inline src/hb-subset-plan.cc:113]: This function and the above are pretty much identical except for the function they call. You should be able to have just one function that takes a function point to the collect function to call.

File: src/hb-subset-plan.cc [hunk 3]
```diff
@@ -68,20 +68,95 @@ _remap_indexes (const hb_set_t *indexes,
...
+				 table_tag,
+				 nullptr,
+				 nullptr,
+				 features,
+				 lookup_indices);
+
+  free (features);
+}
 static inline void
 _gsub_closure_glyphs_lookups_features (hb_face_t *face,
```
Comment (garretrieger) [inline src/hb-subset-plan.cc:146]: This function and _gpos_closure_glyphs_lookups_features are basically identical (except gsub has an additional call to   hb_ot_layout_lookups_substitute_closure). They could be merged into one function by templating the function with the table class (ie. OT:GSUB/OT:GPOS). The tag can be grabbed from the table class via 'tableTag' and then you can put an if (tableTag == HB_OT_TAG_GSUB) hb_ot_layout_lookups_substitute_closure(...

File: test/subset/data/tests/basics.tests [hunk 4]
```diff
@@ -12,6 +12,7 @@ name-ids.txt
 name-languages.txt
 name-legacy.txt
 gids.txt
+layout-features.txt
```
Comment (garretrieger) [inline test/subset/data/tests/basics.tests:16]: Can you also add a test that uses --layout-features=*

**Comment by behdad** [comment]
LGTM w Garret's comments addressed. Thanks.

**Pull request merged**